src/main/scala/firrtl/graph/EulerTour.scala

```diff
@@ -0,0 +1,224 @@
+// See LICENSE for license details.
+
+package firrtl.graph
+
+import scala.collection.mutable
+
+/** Euler Tour companion object */
+object EulerTour {
+  /** Create an Euler Tour of a `DiGraph[T]` */
+  def apply[T](diGraph: DiGraph[T], start: T): EulerTour[Seq[T]] = {
+    val r = mutable.Map[Seq[T], Int]()
+    val e = mutable.ArrayBuffer[Seq[T]]()
+    val h = mutable.ArrayBuffer[Int]()
+
+    def tour(u: T, parent: Seq[T] = Seq(), height: Int = 0): Unit = {
+      val id = parent :+ u
+      if (!r.contains(id)) { r(id) = e.size }
```

Comment:
Seems like a good place for [getOrElseUpdate](http://www.scala-lang.org/api/current/scala/collection/mutable/Map.html#groupBy[K](f:A=>K):scala.collection.immutable.Map[K,Repr])